### ISSUE TITLE
Better documentation for TPUs (e.g., misreporting of WPS)

### ISSUE THREAD

**Issue by kkissmart** [pytorch/fairseq#2327]
#### What is your question?
Can anyone address the problem here?
https://github.com/pytorch/xla/issues/2346

#### Code
export VERSION=nightly
conda activate torch-xla-nightly

python train.py $HOME/pytorch-tutorial-data/wmt18_en_de_bpej32k --save-interval=1 --arch=transformer_vaswani_wmt_en_de_big --attention-dropout=0.1 --no-progress-bar --criterion=label_smoothed_cross_entropy --source-lang=en --lr-scheduler=inverse_sqrt --min-lr 1e-09 --skip-invalid-size-inputs-valid-test --target-lang=de --label-smoothing=0.1 --update-freq=1 --optimizer adam --adam-betas '(0.9, 0.98)' --warmup-init-lr 1e-07 --lr 0.0005 --warmup-updates 4000 --share-all-embeddings --dropout 0.3 --weight-decay 0.0 --valid-subset=valid --max-epoch=25 --tpu --bf16 --max-tokens 4096 --num-batch-buckets 8 --distributed-world-size 8

**Comment by myleott** [comment]
I commented on the attached XLA issue. The reported words per second on TPUs are lower by a factor of `--log-interval`. So in your case you need to multiply the reported wps by 100 (assuming the default log-interval of 100).

**Comment by kkissmart** [comment]
@myleott thanks so much! two more questions

1) does the command look right to you? anything args missed?
2) can you hint any tricks on setting --num-batch-buckets?

Thank you in advance!

**Comment by myleott** [comment]
1. I would recommend removing the `--bf16` flag. While on GPUs the `--fp16` flag improves performance a lot, the speedup is much smaller on TPUs. Just adding `--tpu` should be better to start with.
2. Generally a larger value for `--num-batch-buckets` will make training faster after all the XLA compilations are done. But, larger values will also make training slower at first, since it has to do lots of compiles. I had used 5 in the past, but 8 is probably fine too.

**Comment by taylanbil** [comment]
@myleott , it seems like the underreporting of wps is something that's going to come back, are there plans to fix that? Maybe temporarily multiply w/ `--log-interval` to get a better approximation?

**Comment by kkissmart** [comment]
double check here -- if I run roberta on complete mode, I dont need this --num-batch-buckets mode, since all the tensors will have the same padding (args.num_tokens_sample)
is this correct?
Thanks a lot!

**Comment by kkissmart** [comment]
I tried on similar command on bart (complete mode and no num-batch-buckets args) and logging every update. but the wps is extremely low. Anything missing here?

Thanks @myleott 

2020-07-14 19:33:57 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 19:33:57 | INFO | train_inner | epoch 001:      4 / 183642 loss=15.255, ppl=39116.5, wps=412.4, ups=0.01, wpb=31444, bsz=32, num_updates=4, lr=4e-08, gnorm=2.892, train_wall=76, wall=457
2020-07-14 19:35:16 | INFO | root | NOTE: XLA compilation detected; too many of these can lead to slow training, but we expect a few in the beginning
2020-07-14 19:35:16 | INFO | train_inner | epoch 001:      5 / 183642 loss=15.919, ppl=61972.8, wps=405, ups=0.01, wpb=32036, bsz=32, num_updates=5, lr=5e-08, gnorm=3.671, train_wall=79, wall=536

**Comment by taylanbil** [comment]
every time you log xla tensors (such as loss), there's a context switch. It is strongly recommended to log much less frequently (`--log-interval 100` is certainly a better choice for example). See https://github.com/pytorch/xla/blob/master/TROUBLESHOOTING.md#known-performance-caveats, specifically the bit where it says `The item() operation explicitly asks to evaluate the result. Don't use it unless it's necessary.`

**Comment by kkissmart** [comment]
@taylanbil 
on your tpu fairseq source code, I see this 
            if self.mask_whole_words is not None:
                word_begins_mask = self.mask_whole_words.gather(0, item)
                word_begins_idx = word_begins_mask.nonzero().view(-1)
                sz = len(word_begins_idx)
                words = np.split(word_begins_mask, word_begins_idx)[1:]
                assert len(words) == sz
                word_lens = list(map(len, words))
in the fairseq/data/mask_tokens_dataset.py  code. 

I change the nonzero() into a for loop, but it didnt make any changes.
Anything hint?

**Comment by kkissmart** [comment]
I changed src data in bart with a fixed length, and now the wps is pretty fast ~ 60K tokens per second.

**Comment by zphang** [comment]
What's the currently most recommended dataset-impl for use with TPUs? Particularly in the case of training with pods.

**Comment by myleott** [comment]
> What's the currently most recommended dataset-impl for use with TPUs? Particularly in the case of training with pods.

You usually don't need to set dataset-impl, it's inferred automatically. It's just the data binarization format, has nothing to do with TPUs specifically.

**Comment by zphang** [comment]
It's somewhat orthogonal, but I'm curious about which format you think has the best memory/speed (?) trade-off for a TPU-training workflow. Relatedly, have you found it beneficial to use `num_workers>0` for keeping the TPUs fed with data?

**Comment by salrowili** [comment]
> I changed src data in bart with a fixed length, and now the wps is pretty fast ~ 60K tokens per second.

How I can change the fixed length for BART?
